### PR TITLE
Force dynamic staticity for modules without a guaranteed cmx

### PR DESCRIPTION
**Pull request by riaqn**
A compilation unit can be used at `@ static` (compile-time evaluation) only if a `.cmx` is guaranteed available. `-Ix <dir>` marks a directory where `.cmx` files are guaranteed (unlike plain `-I`), so a module loaded from a plain `-I` directory is forced to `dynamic` regardless of what its `.cmi` records, with a hint explaining why.

`persistent_signature` now carries the unit's mode (`signature * Mode.Value.l`); the `.cmi` still records only `Mode.Staticity.Const.t`. `mode_unit` moves from `Env` to `Persistent_env` (which knows whether a `.cmx` is guaranteed) and is renamed `mode_pers_mod`; for a non-guaranteed `.cmx`, the loaded mode is joined with a `Cmx_not_guaranteed`-hinted `dynamic` to raise its floor.

Tested in `dash-Ix/staticity.ml`: the same static module is rebindable at `@ static` via `-Ix` but not via plain `-I`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: testsuite/tests/dash-Ix/use_static.ml
Comment (jbachurski) [inline testsuite/tests/dash-Ix/use_static.ml:1]: Perhaps you could move the contents of this file to `staticity.ml`? 
It's a bit awkward that `staticity.ml` has no code in it, but always compiles this file anyway.

File: typing/mode.ml [hunk 1]
```diff
@@ -4891,6 +4893,12 @@ module Report = struct
         (print_pp ~definite:false ~capitalize:false)
     | Always_dynamic x ->
       Fmt.fprintf ppf "%t are always dynamic" (print_always_dynamic x)
+    | Cmx_not_guaranteed ->
+      (* CR-someday zqian: this message is specific to the Jane Street build
+         setup and should be made more general. *)
+      Fmt.fprintf ppf
+        "only modules from core libraries and the current library can \
+         currently be static"
```
Comment (jbachurski) [inline typing/mode.ml]: Well, I agree with the CR. I think this will potentially end up quite confusing (what are "core libraries"?). 
In case someone wants to debug this, what about adding a message like `(the .cmx file for X is missing)` (and tracking the module `X` in the hint) mirroring the error message for `Missing_cmi`?

File: typing/mode_hint.mli [hunk 2]
```diff
@@ -120,6 +118,9 @@ type 'd const =
   | Stack_expression : ('l * disallowed) pos const
   | Module_allocated_on_heap : (disallowed * 'r) pos const
   | Always_dynamic : always_dynamic -> ('l * disallowed) neg const
+  | Cmx_not_guaranteed : ('l * disallowed) neg const
+      (** A module is dynamic because no [.cmx] is guaranteed to be available
+          for it, so it cannot be used for compile-time evaluation. *)
```
Comment (jbachurski) [inline typing/mode_hint.mli]: In contrast to `Is_used_in`, there is no invariant to document here, and I think the error message should be sufficiently explanatory for what this case means. 
(I don't like the excessive comments that AI adds describing its initial prompt, especially when it's inconsistent with surrounding code.)

File: typing/mode_hint.mli [hunk 3]
```diff
@@ -120,6 +118,9 @@ type 'd const =
   | Stack_expression : ('l * disallowed) pos const
   | Module_allocated_on_heap : (disallowed * 'r) pos const
   | Always_dynamic : always_dynamic -> ('l * disallowed) neg const
+  | Cmx_not_guaranteed : ('l * disallowed) neg const
```
Comment (jbachurski) [inline typing/mode_hint.mli]: What about tracking the `Global_module.Name.t` here to make the error message more specific?

File: testsuite/tests/dash-Ix/use_static.ml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+(* Rebind the library module [S] at [@ static].  [S]'s cmi records [Static], so
+   this succeeds only when [S] is loaded from a directory where a cmx is
+   guaranteed to be available (i.e. via [-Ix]).  With plain [-I], [S] is forced
+   to [dynamic] and the rebind fails. *)
+module (M @ static) = S
```
Comment (jbachurski) [inline testsuite/tests/dash-Ix/use_static.ml]: The error might be low quality if we have an `open M` and the static-expected use of `M` only occurs later in the file.
(This ties in to the point of adding `global_name`/`Global_module.Name.t` to the hint and error message.)

File: typing/persistent_env.mli [hunk 5]
```diff
@@ -95,6 +95,12 @@ type address =
   | Alocal of Ident.t
   | Adot of address * Types.module_representation * int
 
+(* The mode of a compilation unit: legacy on every axis, with the given
+   staticity. *)
+(* CR-soon zqian: all persistent modules should always be [Static], at which
+   point the [staticity] parameter can be removed. *)
```
Comment (jbachurski) [inline typing/persistent_env.mli:101]: What does this comment mean? Are we going to guarantee `.cmx` files for every library? What about files that are marked with the `@@ dynamic` modality?

File: typing/persistent_env.ml [hunk 6]
```diff
@@ -546,6 +546,26 @@ let check_for_unset_parameters penv global =
            }))
     global.Global_module.hidden_args
 
+(* The mode of a compilation unit: legacy on every axis, with the given
+   staticity. *)
+(* CR-soon zqian: all persistent modules should always be [Static], at which
+   point the [staticity] parameter can be removed. *)
+let mode_pers_mod staticity =
```
Comment (jbachurski) [inline typing/persistent_env.ml]: Are 'compilation unit' (doc comment) and 'persistent module' (CR, function name) synonymous?

File: typing/persistent_env.ml [hunk 7]
```diff
@@ -546,6 +546,26 @@ let check_for_unset_parameters penv global =
            }))
     global.Global_module.hidden_args
 
+(* The mode of a compilation unit: legacy on every axis, with the given
+   staticity. *)
```
Comment (jbachurski) [inline typing/persistent_env.ml]: Why is this comment both in the `.ml` and `.mli`? I think the convention is usually for doc comments to be in the `.mli`, and technical details (or CRs) in the `.ml`.

File: typing/persistent_env.ml [hunk 8]
```diff
@@ -546,6 +546,26 @@ let check_for_unset_parameters penv global =
...
+      uniqueness = Aliased;
+      portability = Nonportable;
+      contention = Uncontended;
+      forkable = Forkable;
+      yielding = Unyielding;
+      statefulness = Stateful;
+      visibility = Read_write;
+      staticity;
+    }
+    ~hint_monadic:hint ~hint_comonadic:hint
```
Comment (jbachurski) [inline typing/persistent_env.ml:553]: Perhaps this could use `{ Value.Const.legacy with staticity }` like in `Typecore.mode_toplevel_expression`?

File: typing/env.ml
Comment (jbachurski) [inline typing/env.ml:3263]: I think it's subtle here that the zapping should never do anything since the underlying `mode` should be a constant. Is there any way to assert that? Otherwise I think this should at least be documented.

File: testsuite/tests/dash-Ix/use_static.ml (hunk 4 above)
Comment (jbachurski) [inline testsuite/tests/dash-Ix/use_static.ml]: There should also be a test to see the hint behaves nicely for example like `S.x` or a function call to a layout-polymorphic `S.foo` (if that's already implemented).

File: typing/persistent_env.ml [hunk 9]
```diff
@@ -686,9 +706,31 @@ and acknowledge_new_pers_name penv check global_name global import =
...
          ~mentioned_by:(Other global_name))
     sign.bound_globals;
+  let pn_sign =
+    let signature, staticity = sign.sign in
+    let mode = Mode.Value.disallow_right (mode_pers_mod staticity) in
+    let mode =
+      match import.imp_visibility with
+      | Visible { cmx_guaranteed = true } ->
+        (* A [.cmx] is guaranteed, so the unit keeps the staticity recorded in
+           its [.cmi]. *)
```
Comment (jbachurski) [inline typing/persistent_env.ml]: I feel like this is comment is unnecessary as it's just reading out the two lines of code.

File: typing/persistent_env.ml [hunk 10]
```diff
@@ -686,9 +706,31 @@ and acknowledge_new_pers_name penv check global_name global import =
...
+        mode
+      | Visible { cmx_guaranteed = false } | Hidden ->
+        (* Without a guaranteed [.cmx], the unit is not available for
+           compile-time evaluation, so its staticity is forced to [Dynamic]
+           regardless of what the [.cmi] claims. The [Cmx_not_guaranteed] hint
+           explains why in any resulting error. *)
```
Comment (jbachurski) [inline typing/persistent_env.ml]: I think it'd be better if the comment tersely explained that this incantation forces the mode to always be `Dynamic`, and we do not return a constant (`Mode.Value.Const.t`) so that we can provide a hint (which is a non-obvious detail IMO).

It might still be worth to have an overarching comment about the point of this code. Perhaps a two-line comment above the `match import.imp_visibility with` line with content like "Compilation units are only available for static evaluation if they have a `.cmx` file, so we force them do be dynamic if we cannot guarantee a `.cmx`".)

**Review by jbachurski — COMMENTED** [review]
I think there's some improvements that can be made around the testing, comments, and the error message. 
The AI comment style doesn't really fit in with the code base and is often unhelpful or redundant.
I'm happy with the core logic.

File: typing/mode.ml (hunk 1 above)
Comment (riaqn) [reply typing/mode.ml]: Ideally the compiler should talk about `-Ix` and build system should re-intepret that into languages about "core libraries".

But currently the build system just forwards whatever is printed by the compiler, therefore the compiler needs to talk about "core libraries". @ccasin thinks our internal users never invoke compiler manually so it's fine.

File: typing/persistent_env.ml (hunk 6 above)
Comment (riaqn) [reply typing/persistent_env.ml]: No. For example, a parameterized module is backed by a single compilation unit (a single cmi/cmx), but multiple instances can co-exists and imported as multiple persistent modules.

File: typing/persistent_env.mli (hunk 5 above)
Comment (riaqn) [reply typing/persistent_env.mli:101]: In near future we want to make all persistent modules to be `static`. When imported, if they are found via `-I` they weakened to `dynamic`.

File: testsuite/tests/dash-Ix/use_static.ml
Comment (riaqn) [reply testsuite/tests/dash-Ix/use_static.ml:1]: Since we will add more tests to `staticity.ml` (as suggested below), it's probably better to keep `staticity.ml` as the list of test directives.

File: typing/mode.ml (hunk 1 above)
Comment (jbachurski) [reply typing/mode.ml]: Hmm, you're right. Perhaps something like `(the .cmx file for X is not marked as guaranteed)`, but you're right there's another detail at play here. 
I think it'd still help to write in the message which module is the problem to hopefully improve the debugging experience.

File: typing/persistent_env.mli (hunk 5 above)
Comment (jbachurski) [reply typing/persistent_env.mli:101]: Hmm, I'm still a bit confused on what would change, then. That sounds like what we're doing now (they are static if the cmx is guaranteed and otherwise dynamic)?

File: typing/persistent_env.ml (hunk 6 above)
Comment (jbachurski) [reply typing/persistent_env.ml]: I see, then it's good you removed the comment as it associated the two ("mode of a compilation unit" on `mode_pers_mod`).

File: testsuite/tests/dash-Ix/use_static.ml (hunk 4 above)
Comment (riaqn) [reply testsuite/tests/dash-Ix/use_static.ml]: good point. Tests added, and indeed the hint are dropped in such cases.

I'm working on a fix that will be a separate PR.

Comment (riaqn) [reply testsuite/tests/dash-Ix/use_static.ml]: Tests added - and indeed the error message drops the hint. I'm working on the fix that will be a separate PR.

File: typing/env.ml
Comment (riaqn) [reply typing/env.ml:3263]: There is `Mode.Value.Guts.check_const` but it requires the mode to be `allowed * allowed`. There is also `check_const_conservative` but it exposes too much internals of the mode solver.

File: typing/persistent_env.mli (hunk 5 above)
Comment (riaqn) [reply typing/persistent_env.mli:101]: currently, if you write a regular `.mli` and compile it, the cmi will claim to be `dynamic`. Only if you write `@@ dynamic` or `@@ static` at the begining of the `.mli` will the generated cmi claim to be `static` (which is further subject to `-Ix`).

Comment (jbachurski) [reply typing/persistent_env.mli:101]: Ahh, I see, so changing the default. Thanks!

File: testsuite/tests/dash-Ix/staticity.ml
Comment (jbachurski) [inline testsuite/tests/dash-Ix/staticity.ml:1]: I'm fine with keeping this as just calling out to different modules (though I don't think it's necessary, but the comment is somewhat useful). 
Could you add a comment at the end explaining that this test file is intentionally empty as it's only a driver?

File: typing/env.ml
Comment (jbachurski) [reply typing/env.ml:3263]: I think a comment would be enough here if this is difficult to assert.

File: typing/persistent_env.ml (hunk 10 above)
Comment (riaqn) [reply typing/persistent_env.ml]: I put the comment about constant at `types.mli` where I think is more appropriate.

**Review by jbachurski — APPROVED** [review]
LGTM
